File: frontend/src/lib/email.ts
import { Resend } from "resend";
import type { BuildItem, Lead } from "@/types";

function formatCents(cents: number): string {
  const hasCents = cents % 100 !== 0;
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: hasCents ? 2 : 0,
    maximumFractionDigits: 2,
  }).format(cents / 100);
}

function wrapperHtml(inner: string): string {
  return `
<!DOCTYPE html>
<html>
<head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"></head>
<body style="margin:0;padding:0;background-color:#0a0a0a;font-family:-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
  <table width="100%" cellpadding="0" cellspacing="0" style="background-color:#0a0a0a;padding:40px 20px;">
    <tr>
      <td align="center">
        <table width="100%" cellpadding="0" cellspacing="0" style="max-width:560px;">
          <tr>
            <td style="padding-bottom:24px;">
              <p style="margin:0;font-size:22px;font-weight:700;color:#f59e0b;">
                &#9881; RideForge
              </p>
            </td>
          </tr>
          ${inner}
          <tr>
            <td style="padding-top:24px;">
              <p style="margin:0;font-size:12px;color:#444;text-align:center;">
                RideForge — customize any car. Manage your account at your RideForge dashboard.
              </p>
            </td>
          </tr>
        </table>
      </td>
    </tr>
  </table>
</body>
</html>`.trim();
}

function getResend(): Resend | null {
  const apiKey = process.env.RESEND_API_KEY;
  if (!apiKey) return null;
  return new Resend(apiKey);
}

export async function sendVerificationEmail(to: string, token: string): Promise<void> {
  const resend = getResend();
  if (!resend) { console.warn("[email] RESEND_API_KEY not set — skipping verification email"); return; }
  const appUrl = process.env.NEXT_PUBLIC_APP_URL || "http://localhost:3000";
  const link = `${appUrl}/verify-email?token=${token}`;
  try {
    await resend.emails.send({
      from: "RideForge <noreply@rideforge.app>",
      to,
      subject: "Verify your RideForge account",
      html: wrapperHtml(`
        <tr><td style="background-color:#111111;border:1px solid #1a1a1a;border-radius:12px;padding:24px;">
          <h1 style="font-size:20px;margin:0 0 12px;color:#f5f5f5;">Verify your email</h1>
          <p style="color:#888;margin:0 0 24px;">Click the button below to verify your email address and activate your account.</p>
          <a href="${link}" style="display:inline-block;background:#f59e0b;color:#000;font-weight:700;padding:12px 28px;border-radius:8px;text-decoration:none;">Verify Email</a>
          <p style="color:#444;font-size:12px;margin-top:24px;">Link expires in 24 hours. If you didn't create a RideForge account, ignore this email.</p>
        </td></tr>`),
    });
  } catch (err) { console.error("[email] Failed to send verification email:", err); }
}

export async function sendPasswordResetEmail(to: string, token: string): Promise<void> {
  const resend = getResend();
  if (!resend) { console.warn("[email] RESEND_API_KEY not set — skipping reset email"); return; }
  const appUrl = process.env.NEXT_PUBLIC_APP_URL || "http://localhost:3000";
  const link = `${appUrl}/reset-password?token=${token}`;
  try {
    await resend.emails.send({
      from: "RideForge <noreply@rideforge.app>",
      to,
      subject: "Reset your RideForge password",
      html: wrapperHtml(`
        <tr><td style="background-color:#111111;border:1px solid #1a1a1a;border-radius:12px;padding:24px;">
          <h1 style="font-size:20px;margin:0 0 12px;color:#f5f5f5;">Reset your password</h1>
          <p style="color:#888;margin:0 0 24px;">Click below to set a new password. This link expires in 1 hour.</p>
          <a href="${link}" style="display:inline-block;background:#f59e0b;color:#000;font-weight:700;padding:12px 28px;border-radius:8px;text-decoration:none;">Reset Password</a>
          <p style="color:#444;font-size:12px;margin-top:24px;">If you didn't request this, you can safely ignore this email.</p>
        </td></tr>`),
    });
  } catch (err) { console.error("[email] Failed to send reset email:", err); }
}

export async function sendOrderConfirmationEmail(
  to: string,
  order: { id: string; items: BuildItem[]; subtotal_cents: number; shipping_cents: number; total_cents: number }
): Promise<void> {
  const resend = getResend();
  if (!resend) { console.warn("[email] RESEND_API_KEY not set — skipping order email"); return; }

  const rows = order.items
    .map(
      (it) => `
      <tr>
        <td style="padding:8px 0;color:#e4e4e7;font-size:14px;">${it.name} <span style="color:#71717a;">×${it.qty}</span></td>
        <td style="padding:8px 0;color:#e4e4e7;font-size:14px;text-align:right;">${formatCents(it.price_cents * it.qty)}</td>
      </tr>`
    )
    .join("");

  try {
    await resend.emails.send({
      from: "RideForge Orders <orders@rideforge.app>",
      to,
      subject: `Order confirmed — #${order.id.slice(0, 8).toUpperCase()}`,
      html: wrapperHtml(`
        <tr><td style="background-color:#111111;border:1px solid #1a1a1a;border-radius:12px;padding:24px;">
          <h1 style="font-size:20px;margin:0 0 4px;color:#f5f5f5;">Order confirmed 🎉</h1>
          <p style="color:#888;margin:0 0 20px;font-size:13px;">Order #${order.id.slice(0, 8).toUpperCase()}</p>
          <table width="100%" cellpadding="0" cellspacing="0" style="border-top:1px solid #27272a;padding-top:12px;">
            ${rows}
            <tr><td style="padding-top:12px;color:#71717a;font-size:13px;">Subtotal</td><td style="padding-top:12px;color:#e4e4e7;font-size:13px;text-align:right;">${formatCents(order.subtotal_cents)}</td></tr>
            <tr><td style="color:#71717a;font-size:13px;">Shipping</td><td style="color:#e4e4e7;font-size:13px;text-align:right;">${order.shipping_cents === 0 ? "Free" : formatCents(order.shipping_cents)}</td></tr>
            <tr><td style="padding-top:8px;border-top:1px solid #27272a;color:#f5f5f5;font-weight:700;">Total</td><td style="padding-top:8px;border-top:1px solid #27272a;color:#f59e0b;font-weight:700;text-align:right;">${formatCents(order.total_cents)}</td></tr>
          </table>
        </td></tr>`),
    });
  } catch (err) { console.error("[email] Failed to send order confirmation:", err); }
}

export async function sendLeadNotificationEmail(to: string, lead: Lead): Promise<void> {
  const resend = getResend();
  if (!resend) { console.warn("[email] RESEND_API_KEY not set — skipping lead notification"); return; }

  const carLine = [lead.year, lead.make, lead.model].filter(Boolean).join(" ") || "Unspecified vehicle";
  const budgetLine = lead.budget_cents ? formatCents(lead.budget_cents) : "Not specified";

  try {
    await resend.emails.send({
      from: "RideForge Leads <leads@rideforge.app>",
      to,
      subject: `New install lead near ${lead.zip}${lead.category ? ` — ${lead.category}` : ""}`,
      html: wrapperHtml(`
        <tr><td style="background-color:#111111;border:1px solid #1a1a1a;border-radius:12px;padding:24px;">
          <p style="margin:0 0 12px;font-size:13px;color:#f59e0b;text-transform:uppercase;letter-spacing:0.05em;font-weight:600;">New Lead</p>
          <h1 style="font-size:20px;margin:0 0 16px;color:#f5f5f5;">${carLine}</h1>
          <p style="margin:4px 0;color:#888;font-size:14px;">ZIP: <strong style="color:#f5f5f5;">${lead.zip}</strong></p>
          <p style="margin:4px 0;color:#888;font-size:14px;">Category: <strong style="color:#f5f5f5;">${lead.category ?? "Any"}</strong></p>
          <p style="margin:4px 0;color:#888;font-size:14px;">Budget: <strong style="color:#f5f5f5;">${budgetLine}</strong></p>
          ${lead.notes ? `<p style="margin:12px 0 0;color:#a1a1aa;font-size:13px;">"${lead.notes}"</p>` : ""}
          <p style="margin-top:20px;color:#444;font-size:12px;">Sign in to your Shop Portal to claim this lead and view full contact details.</p>
        </td></tr>`),
    });
  } catch (err) { console.error("[email] Failed to send lead notification:", err); }
}

File: backend/app/services/email.py
import logging
from datetime import datetime, timezone

from sendgrid import SendGridAPIClient
from sendgrid.helpers.mail import Mail

from app.config import get_settings

logger = logging.getLogger(__name__)
settings = get_settings()


def _send_email(to_email: str, subject: str, html_content: str) -> bool:
    """Send an email via SendGrid. Returns True on success."""
    if not settings.sendgrid_api_key:
        logger.warning("SendGrid API key not configured, skipping email to %s", to_email)
        return False

    message = Mail(
        from_email=settings.from_email,
        to_emails=to_email,
        subject=subject,
        html_content=html_content,
    )
    try:
        sg = SendGridAPIClient(settings.sendgrid_api_key)
        response = sg.send(message)
        if response.status_code in (200, 202):
            logger.info("Email sent to %s: %s", to_email, subject)
            return True
        else:
            logger.error("SendGrid returned status %s for email to %s", response.status_code, to_email)
            return False
    except Exception as e:
        logger.error("Failed to send email to %s: %s", to_email, e, exc_info=True)
        return False


def send_verification_email(to_email: str, token: str) -> bool:
    verify_url = f"{settings.frontend_url}/verify-email?token={token}"
    html = f"""
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Verify your email</title>
    </head>
    <body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
                 background: #f4f4f5; margin: 0; padding: 40px 20px;">
      <div style="max-width: 480px; margin: 0 auto; background: #fff;
                  border-radius: 12px; overflow: hidden; box-shadow: 0 2px 8px rgba(0,0,0,0.08);">
        <div style="background: #18181b; padding: 32px; text-align: center;">
          <h1 style="color: #fff; margin: 0; font-size: 24px; letter-spacing: -0.5px;">Sniper</h1>
          <p style="color: #a1a1aa; margin: 8px 0 0; font-size: 14px;">Enthusiast Car Auction Alerts</p>
        </div>
        <div style="padding: 40px 32px;">
          <h2 style="color: #18181b; margin: 0 0 16px; font-size: 20px;">Verify your email address</h2>
          <p style="color: #52525b; line-height: 1.6; margin: 0 0 32px;">
            Thanks for signing up! Click the button below to verify your email address
            and start getting auction alerts.
          </p>
          <a href="{verify_url}"
             style="display: block; background: #2563eb; color: #fff; text-decoration: none;
                    padding: 14px 24px; border-radius: 8px; text-align: center;
                    font-weight: 600; font-size: 16px; margin-bottom: 24px;">
            Verify Email Address
          </a>
          <p style="color: #a1a1aa; font-size: 13px; margin: 0; text-align: center;">
            This link expires in 24 hours. If you didn't create an account, ignore this email.
          </p>
        </div>
      </div>
    </body>
    </html>
    """
    return _send_email(to_email, "Verify your Sniper account", html)


def send_password_reset_email(to_email: str, token: str) -> bool:
    reset_url = f"{settings.frontend_url}/reset-password?token={token}"
    html = f"""
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Reset your password</title>
    </head>
    <body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
                 background: #f4f4f5; margin: 0; padding: 40px 20px;">
      <div style="max-width: 480px; margin: 0 auto; background: #fff;
                  border-radius: 12px; overflow: hidden; box-shadow: 0 2px 8px rgba(0,0,0,0.08);">
        <div style="background: #18181b; padding: 32px; text-align: center;">
          <h1 style="color: #fff; margin: 0; font-size: 24px; letter-spacing: -0.5px;">Sniper</h1>
          <p style="color: #a1a1aa; margin: 8px 0 0; font-size: 14px;">Enthusiast Car Auction Alerts</p>
        </div>
        <div style="padding: 40px 32px;">
          <h2 style="color: #18181b; margin: 0 0 16px; font-size: 20px;">Reset your password</h2>
          <p style="color: #52525b; line-height: 1.6; margin: 0 0 32px;">
            We received a request to reset your password. Click below to choose a new one.
            If you didn't request this, you can safely ignore this email.
          </p>
          <a href="{reset_url}"
             style="display: block; background: #2563eb; color: #fff; text-decoration: none;
                    padding: 14px 24px; border-radius: 8px; text-align: center;
                    font-weight: 600; font-size: 16px; margin-bottom: 24px;">
            Reset Password
          </a>
          <p style="color: #a1a1aa; font-size: 13px; margin: 0; text-align: center;">
            This link expires in 24 hours.
          </p>
        </div>
      </div>
    </body>
    </html>
    """
    return _send_email(to_email, "Reset your Sniper password", html)


def send_alert_email(to_email: str, alert_name: str, listings: list[dict]) -> bool:
    """Send a multi-listing alert digest email."""
    if not listings:
        return False

    listing_cards = ""
    for listing in listings[:10]:  # cap at 10 per email
        image_url = listing.get("images", [None])[0] or ""
        title = listing.get("title", "Unknown")
        year = listing.get("year", "")
        mileage = listing.get("mileage")
        price = listing.get("price")
        source = listing.get("source", "").replace("_", " ").title()
        url = listing.get("url", "#")
        auction_end = listing.get("auction_end")

        mileage_str = f"{mileage:,} mi" if mileage else "N/A"
        price_str = f"${price:,.0f}" if price else "No reserve / TBD"

        end_str = ""
        if auction_end:
            if isinstance(auction_end, str):
                end_str = auction_end
            elif hasattr(auction_end, "strftime"):
                end_str = auction_end.strftime("%b %d, %Y %I:%M %p UTC")

        image_html = ""
        if image_url:
            image_html = f"""
            <a href="{url}" style="display: block; margin-bottom: 16px;">
              <img src="{image_url}" alt="{title}"
                   style="width: 100%; height: 200px; object-fit: cover; border-radius: 8px;" />
            </a>
            """

        listing_cards += f"""
        <div style="background: #fff; border: 1px solid #e4e4e7; border-radius: 12px;
                    padding: 20px; margin-bottom: 20px;">
          {image_html}
          <div style="display: flex; align-items: center; margin-bottom: 10px;">
            <span style="background: #18181b; color: #fff; font-size: 11px; font-weight: 600;
                         padding: 3px 8px; border-radius: 4px; text-transform: uppercase;
                         letter-spacing: 0.5px;">{source}</span>
          </div>
          <h3 style="margin: 0 0 8px; font-size: 17px; color: #18181b; line-height: 1.3;">
            <a href="{url}" style="color: #18181b; text-decoration: none;">{title}</a>
          </h3>
          <div style="color: #52525b; font-size: 14px; margin-bottom: 12px;">
            <span style="margin-right: 16px;">Year: <strong>{year}</strong></span>
            <span style="margin-right: 16px;">Mileage: <strong>{mileage_str}</strong></span>
            <span>Current bid: <strong style="color: #16a34a;">{price_str}</strong></span>
          </div>
          {"<div style='color: #dc2626; font-size: 13px; margin-bottom: 12px;'>&#9200; Ends: " + end_str + "</div>" if end_str else ""}
          <a href="{url}"
             style="display: inline-block; background: #2563eb; color: #fff; text-decoration: none;
                    padding: 10px 20px; border-radius: 6px; font-weight: 600; font-size: 14px;">
            View Auction &rarr;
          </a>
        </div>
        """

    manage_url = f"{settings.frontend_url}/alerts"
    count = len(listings)
    html = f"""
    <!DOCTYPE html>
    <html>
    <head>
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>New auction matches for "{alert_name}"</title>
    </head>
    <body style="font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
                 background: #f4f4f5; margin: 0; padding: 40px 20px;">
      <div style="max-width: 600px; margin: 0 auto;">
        <div style="background: #18181b; border-radius: 12px 12px 0 0; padding: 28px 32px;">
          <h1 style="color: #fff; margin: 0; font-size: 22px; letter-spacing: -0.5px;">
            Sniper <span style="color: #60a5fa;">&#x2022;</span> {count} new match{"es" if count != 1 else ""}
          </h1>
          <p style="color: #a1a1aa; margin: 6px 0 0; font-size: 14px;">
            Alert: <strong style="color: #e4e4e7;">{alert_name}</strong>
          </p>
        </div>
        <div style="background: #f9f9fb; padding: 24px 32px;">
          {listing_cards}
          <div style="text-align: center; padding-top: 16px; border-top: 1px solid #e4e4e7;">
            <a href="{manage_url}" style="color: #71717a; font-size: 13px; text-decoration: none;">
              Manage your alerts
            </a>
          </div>
        </div>
      </div>
    </body>
    </html>
    """

    subject = f"[Sniper] {count} new match{'es' if count != 1 else ''} for \"{alert_name}\""
    return _send_email(to_email, subject, html)
